### DISCUSSION TITLE
A<Action<T>>.That.ShouldBeEquivalentTo

### DISCUSSION THREAD

**Discussion by RLittlesII** [FakeItEasy/FakeItEasy#2041] [General]
I have an API surface of

`IService.Method(String name, Action<T> builder)`

I want to test that the `builder` is constructed correctly.

When I use `A<Action<T>>.That.ShouldBeEquiivalentTo(equivalentBuilder)` it fails to assert the equivalency.

If I provide an action that mirrors what I have in my system under tests, it still fails to see the two actions as the same.

i.e. `A.CallTo(() => service.Method(A<string>._, (thing) => thing.With("value")))`

I am wondering if there is any guidance for how to achieve this with Fake It Easy.

I have looked at the inter-webs and found the following.  Just making sure it's a good direction before I jump down the rabbit hole.

- https://stackoverflow.com/questions/19974057/how-to-fake-an-action-with-fakeiteasy
- https://stackoverflow.com/questions/6701041/compare-delegates-actiont
- https://stackoverflow.com/a/23216674

**Comment by blairconrad** [comment]
Hi, @RLittlesII. Thanks for asking.

First, comparing actions is very difficult. The `Equals` method is reference-based. My general advice would be, in descending order

1. if you have access to the actual action that's expected, do a reference comparison via `IsSameAs`
2. capture the action and apply it to an object and then see if it has the desired effect (assuming this is something you can observe). For example, if the action sets a property value on the object, check that
3. try the advice at https://stackoverflow.com/questions/6701041/compare-delegates-actiont to compare the actual contents, but I worry that it's too brittle; _any_ difference in implementation would break the test, even if the effect were the same

I gotta ask, though. 

> When I use `A<Action<T>>.That.ShouldBeEquiivalentTo(equivalentBuilder)` it fails to assert the equivalency.

What's `ShouldBeEquiivalentTo`? I can't find it anywhere (nor even `ShouldBeEquivalentTo`).

**Reply by RLittlesII** [reply]
Thanks for the quick reply!

I figured you'd tell me it was difficult.

On the topic of brittle, the way we are consuming this Fake it Easy in this instance seems brittle given we are asserting method internals instead of method returns.

Option 2 would require me to change my API surface slightly for `<T>` in my case.  Might be worth it if we determine this level of fidelity is required.

> What's ShouldBeEquiivalentTo? I can't find it anywhere (nor even ShouldBeEquivalentTo).

Glad you asked this question.  Seems we wrote an internal extension on our project by this name, I didn't check before I asked the question.  😏 

```csharp
public static T ShouldBeEquivalentTo<T>(this IArgumentConstraintManager<T> constraintManager, T expectedValue) => constraintManager.Matches(actual => AssertEquivalentTo(actual, expectedValue, (config => config)));

...

private static bool AssertEquivalentTo<T>(T actual, T expected, Func<EquivalencyAssertionOptions<T>, EquivalencyAssertionOptions<T>> config)
{
       actual.Should()
             .BeEquivalentTo(expected, config);

       return true;
}
```
